### DISCUSSION TITLE
Beginners first step using Sky()

### DISCUSSION THREAD

**Discussion by wallygit** [pokepetter/ursina#306] [Q&A]
Hello,
I tried the following code:
```python
from ursina import *
app = Ursina()
window.title = "ursina"
skybox_image = load_texture("sky_sunset.jpg")
Sky(texture=skybox_image)
app.run()
```
expecte to see:
![sky_sunset](https://user-images.githubusercontent.com/4344190/139841281-9ba8c2e0-ec12-4f63-a6a0-139680200a4e.jpg)

but I get:
![Screenshot_20211102_122727](https://user-images.githubusercontent.com/4344190/139841417-5b76b2d2-b446-48f2-806c-c7f53c638295.png)
terminal output:
/playground/ursina/sky_02.py
/myvenv/lib/python3.8/site-packages/ursina
asset_folder: /playground/ursina
blender_paths:
{'default': '/usr/bin/blender'}
OS: posix
screen resolution: (1600, 900)
Known pipe types:
  glxGraphicsPipe
(4 aux display modules not yet loaded.)
size; LVector2f(1280, 720)
render mode: default
:pnmimage:png(warning): iCCP: known incorrect sRGB profile
development mode: True
application successfully started
:display:gsg:glgsg(error): An error occurred while compiling GLSL vertex shader created-shader:
created-shader:1(10): error: GLSL 1.40 is not supported. Supported versions are: 1.10, 1.20, 1.30, 1.00 ES, and 3.00 ES
:display:gsg:glgsg(error): An error occurred while compiling GLSL fragment shader created-shader:
created-shader:2(10): error: GLSL 1.40 is not supported. Supported versions are: 1.10, 1.20, 1.30, 1.00 ES, and 3.00 ES
no filter quad
changed aspect ratio: 1.778 -> 1.333

OS: Kubuntu 20.04.1

how can I do it better and where is the fault?
Thanks

**Comment by pokepetter** [comment]
Looks like that GLSL version isn't supported by you hardware, so it didn't compile. I lowered it to version 130, so update and try again: `pip3 install https://github.com/pokepetter/ursina/archive/master.zip --upgrade`. You're likely to run into issues with other shaders too, and there might not be a fix for that.

**Reply by pokepetter** [reply]
Also, the sky image doesn't actually loop horizontally, but that's my bad :P

**Comment by wallygit** [comment]
Hi Petter,
thanks for reply. I have tried to follow your instructions and get the following terminal output.
nothing has changed when testing the script.
(as you can see I even have problems formatting my posts correctly, sorry for that)

terminal output of pip3 install ...  :

<(ursenv) op@wilk:~/WORK/PROJECTS/Python/00_PythonProjects/00_Ursina$ pip3 install https://github.com/pokepetter/ursina/archive/master.zip --upgrade
Collecting https://github.com/pokepetter/ursina/archive/master.zip
  Downloading https://github.com/pokepetter/ursina/archive/master.zip
     - 8.2 MB 932 kB/s
Requirement already satisfied, skipping upgrade: panda3d in ./ursenv/lib/python3.8/site-packages (from ursina==4.0.0) (1.10.10)
Requirement already satisfied, skipping upgrade: panda3d-gltf in ./ursenv/lib/python3.8/site-packages (from ursina==4.0.0) (0.13)
Requirement already satisfied, skipping upgrade: pillow in ./ursenv/lib/python3.8/site-packages (from ursina==4.0.0) (8.4.0)
Requirement already satisfied, skipping upgrade: pyperclip in ./ursenv/lib/python3.8/site-packages (from ursina==4.0.0) (1.8.2)
Requirement already satisfied, skipping upgrade: screeninfo in ./ursenv/lib/python3.8/site-packages (from ursina==4.0.0) (0.7)
Requirement already satisfied, skipping upgrade: panda3d-simplepbr>=0.6 in ./ursenv/lib/python3.8/site-packages (from panda3d-gltf->ursina==4.0.0) (0.9)
Building wheels for collected packages: ursina
  Building wheel for ursina (setup.py) ... error
  ERROR: Command errored out with exit status 1:
   command: /home/op/WORK/PROJECTS/Python/00_PythonProjects/00_Ursina/ursenv/bin/python3 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-req-build-fmpefo1l/setup.py'"'"'; __file__='"'"'/tmp/pip-req-build-fmpefo1l/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /tmp/pip-wheel-l_3lmwl3
       cwd: /tmp/pip-req-build-fmpefo1l/
  Complete output (6 lines):
  usage: setup.py [global_opts] cmd1 [cmd1_opts] [cmd2 [cmd2_opts] ...]
     or: setup.py --help [cmd1 cmd2 ...]
     or: setup.py --help-commands
     or: setup.py cmd --help
  
  error: invalid command 'bdist_wheel'
  ----------------------------------------
  ERROR: Failed building wheel for ursina
  Running setup.py clean for ursina
Failed to build ursina
Installing collected packages: ursina
  Attempting uninstall: ursina
    Found existing installation: ursina 4.0.0
    Uninstalling ursina-4.0.0:
      Successfully uninstalled ursina-4.0.0
    Running setup.py install for ursina ... done
Successfully installed ursina-4.0.0
(ursenv) op@wilk:~/WORK/PROJECTS/Python/00_PythonProjects/00_Ursina$ >

**Reply by pokepetter** [reply]
Oh, that's right, you're on linux. do `pip3 install git+https://github.com/pokepetter/ursina.git` then. This is covered in the installation guide by the way.

**Comment by wallygit** [comment]
some other more sophisticated sample codes work quite well. Could it be that the size of the images or window/screen need to be adjusted to each other?

**Comment by wallygit** [comment]
Petter,
maybe I should reinstall everything again.
Tomorrow I will try again.

terminal output:
`(ursenv) op@wilk:~/WORK/PROJECTS/Python/00_PythonProjects/00_Ursina$ pip install git+https://github.com/pokepetter/ursina.git
Collecting git+https://github.com/pokepetter/ursina.git
  Cloning https://github.com/pokepetter/ursina.git to /tmp/pip-req-build-1vuqzxnw
  Running command git clone -q https://github.com/pokepetter/ursina.git /tmp/pip-req-build-1vuqzxnw
Requirement already satisfied (use --upgrade to upgrade): ursina==4.0.0 from git+https://github.com/pokepetter/ursina.git in ./ursenv/lib/python3.8/site-packages
Requirement already satisfied: panda3d in ./ursenv/lib/python3.8/site-packages (from ursina==4.0.0) (1.10.10)
Requirement already satisfied: panda3d-gltf in ./ursenv/lib/python3.8/site-packages (from ursina==4.0.0) (0.13)
Requirement already satisfied: pillow in ./ursenv/lib/python3.8/site-packages (from ursina==4.0.0) (8.4.0)
Requirement already satisfied: pyperclip in ./ursenv/lib/python3.8/site-packages (from ursina==4.0.0) (1.8.2)
Requirement already satisfied: screeninfo in ./ursenv/lib/python3.8/site-packages (from ursina==4.0.0) (0.7)
Requirement already satisfied: panda3d-simplepbr>=0.6 in ./ursenv/lib/python3.8/site-packages (from panda3d-gltf->ursina==4.0.0) (0.9)
Building wheels for collected packages: ursina
  Building wheel for ursina (setup.py) ... error
  ERROR: Command errored out with exit status 1:
   command: /home/op/WORK/PROJECTS/Python/00_PythonProjects/00_Ursina/ursenv/bin/python3 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-req-build-1vuqzxnw/setup.py'"'"'; __file__='"'"'/tmp/pip-req-build-1vuqzxnw/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /tmp/pip-wheel-olra4z6r
       cwd: /tmp/pip-req-build-1vuqzxnw/
  Complete output (6 lines):
  usage: setup.py [global_opts] cmd1 [cmd1_opts] [cmd2 [cmd2_opts] ...]
     or: setup.py --help [cmd1 cmd2 ...]
     or: setup.py --help-commands
     or: setup.py cmd --help
  
  error: invalid command 'bdist_wheel'
  ----------------------------------------
  ERROR: Failed building wheel for ursina
  Running setup.py clean for ursina
Failed to build ursina`

**Comment by wallygit** [comment]
Hi Petter,
does the ursina code at the very top work for you?
I also have problems with other 3D graphs on my computer. 
e.g. Geogebra 3D "Surface(x,y) = sin( x^2 - y^2)".
The screen flickers and is unusable.
Maybe too weak graphic devices or wrong setup.
...
In the meantime i tried it on another computer with different hardware. 3D rendering works here, but the ursina code does not work either and shows the same result

**Reply by pokepetter** [reply]
Yes it works for me.
I'm confused when you say 3d rendering works on the second computer when the result you posted is an issue with the 3d rendering.
If only the default shader is working, try setting the shader to None. The Sky is supposed to ignore shadows and fog, but if you don use them, you can use the default shader.

**Comment by wallygit** [comment]
Petter,
I am sorry if the statement was unclear.
On PC1 there are problems with e.g. Geogebra 3D and Ursina, on PC2 the rendering generally works e.g. Geogebra 3D but also not Ursina. So I thought it is not because of the hardware. 

> Yes it works for me.

which Ursina version ?
> try setting the shader to None.

OK, I'm trying to figure out how to do this.

**Reply by pokepetter** [reply]
`Entity(model='cube', shader=None)`

and since Sky is an Entity:
`Sky(shader=None)`

**Comment by wallygit** [comment]
the window opens briefly and then it collapses along with the output below.
I think I give up. It took me too much time to get used to it. 
Thanks anyway for your effort.
code:
```python
from ursina import *
from ursina import shader

app = Ursina()
window.title = "ursina"
skybox_image = load_texture("sky_sunset.jpg")
Sky(texture=skybox_image, shader="None")
app.run()
```

terminal output:
(ursenv) op@wilk:~/WORK/PROJECTS/Python/00_PythonProjects/00_Ursina$ /home/op/WORK/PROJECTS/Python/00_PythonProjects/00_Ursina/ursenv/bin/python /home/op/WORK/PROJECTS/Python/00_PythonProjects/00_Ursina/sky_02.py
package_folder: /home/op/WORK/PROJECTS/Python/00_PythonProjects/00_Ursina/ursenv/lib/python3.8/site-packages/ursina
asset_folder: /home/op/WORK/PROJECTS/Python/00_PythonProjects/00_Ursina
OS: posix
screen resolution: (1600, 900)
Known pipe types:
  glxGraphicsPipe
(4 aux display modules not yet loaded.)
:pnmimage:png(warning): iCCP: known incorrect sRGB profile
info: development mode: True
application successfully started
Traceback (most recent call last):
  File "/home/op/WORK/PROJECTS/Python/00_PythonProjects/00_Ursina/ursenv/lib/python3.8/site-packages/ursina/main.py", line 143, in _update
    entity.update()
  File "/home/op/WORK/PROJECTS/Python/00_PythonProjects/00_Ursina/ursenv/lib/python3.8/site-packages/ursina/prefabs/sky.py", line 21, in update
    self.world_position = camera.world_position
  File "/home/op/WORK/PROJECTS/Python/00_PythonProjects/00_Ursina/ursenv/lib/python3.8/site-packages/ursina/entity.py", line 261, in __setattr__
    elif hasattr(self, '_shader') and self.shader and name in self._shader.default_input:
AttributeError: 'str' object has no attribute 'default_input'
:task(error): Exception occurred in PythonTask update
Traceback (most recent call last):
  File "/home/op/WORK/PROJECTS/Python/00_PythonProjects/00_Ursina/sky_02.py", line 15, in <module>
    app.run()
  File "/home/op/WORK/PROJECTS/Python/00_PythonProjects/00_Ursina/ursenv/lib/python3.8/site-packages/ursina/main.py", line 239, in run
    super().run()
  File "/home/op/WORK/PROJECTS/Python/00_PythonProjects/00_Ursina/ursenv/lib/python3.8/site-packages/direct/showbase/ShowBase.py", line 3325, in run
    self.taskMgr.run()
  File "/home/op/WORK/PROJECTS/Python/00_PythonProjects/00_Ursina/ursenv/lib/python3.8/site-packages/direct/task/Task.py", line 546, in run
    self.step()
  File "/home/op/WORK/PROJECTS/Python/00_PythonProjects/00_Ursina/ursenv/lib/python3.8/site-packages/direct/task/Task.py", line 500, in step
    self.mgr.poll()
  File "/home/op/WORK/PROJECTS/Python/00_PythonProjects/00_Ursina/ursenv/lib/python3.8/site-packages/ursina/main.py", line 143, in _update
    entity.update()
  File "/home/op/WORK/PROJECTS/Python/00_PythonProjects/00_Ursina/ursenv/lib/python3.8/site-packages/ursina/prefabs/sky.py", line 21, in update
    self.world_position = camera.world_position
  File "/home/op/WORK/PROJECTS/Python/00_PythonProjects/00_Ursina/ursenv/lib/python3.8/site-packages/ursina/entity.py", line 261, in __setattr__
    elif hasattr(self, '_shader') and self.shader and name in self._shader.default_input:
AttributeError: 'str' object has no attribute 'default_input'

**Reply by pokepetter** [reply]
Compare your code to mine, it's not the same

**Comment by wallygit** [comment]
Yes, sorry:
```python
from ursina import *
app = Ursina()
window.title = "ursina"
skybox_image = load_texture("sky_sunset.jpg")
Sky(texture=skybox_image, shader=None)
app.run()
```
but nothing has changed. The rendering looks the same as at the beginning.

**Reply by CiSienx** [reply]
did you found solution?

**Reply by wallygit** [reply]
no, I have not found a solution and have given up on Ursina :(

**Reply by a2cy** [reply]
does the default sky texture work for you?

**Reply by wallygit** [reply]
I'm really sorry, but I don't remember the details. At the time I spoke to a developer and he couldn't help either. 
Do you have the same problem?
If you can solve it please drop me a short message.
I thought after a year there were enough updates and users who had the same problem and it was solved. So maybe it's not Ursina (?)
I use Kubuntu 20.0.4

**Reply by a2cy** [reply]
I've used Kubuntu and I had no problems but as petter said the texture doesn't display properly because it doesn't loop and has the wrong dimensions.

**Reply by wallygit** [reply]
what are the correct dimensions and how to make it looping ?

**Reply by a2cy** [reply]
you need to edit the texture file

**Reply by a2cy** [reply]
petter updated the texture should work now

**Comment by pokepetter** [comment]
I think the main issue is that you assume it's wrong when it's in fact supposed to look like that. Keep in mind the texture on Sky will cover the whole sky, not just the area you see currently. If you for example rotate the camera, you'll see that it makes sense. The texture doesn't loop properly, but that's a different issue.

**Reply by CiSienx** [reply]
yeah you are right I actually solved my issue by changing entity model to cube.

**Reply by wallygit** [reply]
> I think the main issue is that you assume it's wrong when it's in fact supposed to look like that. Keep in mind the texture on Sky will cover the whole sky, not just the area you see currently. If you for example rotate the camera, you'll see that it makes sense. The texture doesn't loop properly, but that's a different issue.

This seems like a very smart thought to me and makes sense.

**Comment by mahdi05tazwar** [comment]
I've found a solution. You need to use a skybox image (just get one from google). (JPG format is preferred)
